homeassistant/components/notify/xmpp.py

```diff
@@ -93,20 +103,204 @@ def __init__(self):
+
+            upload a file with Jabber XEP_0363 from a remote URL or a local
+            file path and return a URL of that file.
+            """
+            if data.get(ATTR_URL):
+                # send a file from an URL
+                url = data.get(ATTR_URL)
+                _LOGGER.info('getting file from %s', url)
+
+                # result = await loop.run_in_executor(None, requests.get, url)
+                if data.get(ATTR_VERIFY, True):
+                    # if True or not set
+                    result = await hass.async_add_executor_job(requests.get,
+                                                               url)
+                else:
+                    def get_insecure_url(url):
+                        return requests.get(url, verify=False)
+                    result = await hass.async_add_executor_job(
+                        get_insecure_url, url)
+
+                if result.status_code >= 400:
+                    _LOGGER.error("could not load file from %s", url)
+                    return
+
+                length = len(result.content)
+                # we need a file extension, the upload server needs a
+                # filename, if none is provided, through the path
+                # we guess the extension
+                if not data.get(ATTR_PATH):
+                    extension = self.get_extension(
+                        result.headers['Content-Type'])
+                    _LOGGER.debug("got %s extension", extension)
+                filename = data.get(ATTR_PATH) if data.get(ATTR_PATH) \
+                    else "upload"+extension
+                url = await self['xep_0363'].upload_file(
+                    filename,
+                    # size=int(result.headers['Content-Length']),
+                    size=length,
+                    input_file=result.content,
+                    content_type=result.headers['Content-Type'])
+            elif data.get(ATTR_PATH):
+                # send message from local path
+                filename = data.get(ATTR_PATH) if data else None
+                _LOGGER.info('Uploading file %s ...', filename)
+                url = await self['xep_0363'].upload_file(filename)
+            else:
+                _LOGGER.error("no path or URL found for image")
+
+            _LOGGER.info('Upload success!')
+            return url
+
+        def send_text_message(self):
+            """Send a text only message to a room or a recipient."""
             if room:
                 _LOGGER.debug("Joining room %s", room)
                 self.plugin['xep_0045'].join_muc(room, sender, wait=True)
                 self.send_message(mto=room, mbody=message, mtype='groupchat')
             else:
+                _LOGGER.debug("message to %s", recipient)
                 self.send_message(mto=recipient, mbody=message, mtype='chat')
-            self.disconnect(wait=True)
+
+        # pylint: disable=no-self-use
+        def get_extension(self, content_type):
```

Comment:
This looks like a task for `python-magic` or a similar module.

Reply:
I'm actually trying to do the opposite of `magic`. 

A HTTP request answers with a special header `Content-Type`. I use that to 'guess' a file extension that can be used. We don't get a filename, except if specified with `path`, so the upload has to be made to a 'fictional' file name. This is `upload.EXTENSION`. We can do it without, but then a mobile client such as Conversations, does not know what to use to open it. That's why we 'guess' an appropriate extension.
This is not used when a local file is specified...

Reply:
Can't the `mimetypes` module do what you want? it has a function to guess the filetype based on a mime type…

```
In [2]: mimetypes.guess_extension('text/calendar')                                                                                                                                                    
Out[2]: '.ics'
```

Reply:
YES! thank you! I'll use that.
there are some differences and unavailable filetypes, but I guess we can live with that:
```
mimetypes  vs  current_dict
None .aac
None .azw
.a .bin
None .bmp
None .bz
None .bz2
None .epub
None .ico
.jpe .jpg
.mid .midi
None .midi
.m1v .mpeg
None .mpkg
None .otf
.pot .ppt
None .rar
None .ts
None .ttf
.bat .txt
None .wav
None .weba
None .webp
None .woff
None .woff2
.xlb .xls
.rdf .xml
None .3g2
```

Reply:
@flowolf Parts of your None-matches in `mimetypes` occurs because the content-type you have in your list does not match `/etc/mime.types`. Ie `application/rar` vs `application/x-rar-compressed`. `.mid` vs `.midi` can be (in most cases) attacked by `guess_all_extensions` and choossing the longest.